zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,454 @@
+		 * alloc data there.
+		 */
+		if (allocation_error_code == 0)	{
+			new_allocation->allocation_ptr = ptr;
+			new_allocation->allocation_size = blck_size;
+		}
+	}
+
+	/* If ptr is NULL it means we did not allocate anything and we should
+	 * break execution here
+	 */
+	if (ptr == NULL)
+		return ptr;
+
+	/* Now we need to iterate over mapped physical memory check if we
+	 * already have this specific
+	 */
+	struct list_item *iterator;
+	bool space_already_mapped = false;
+
+	list_for_item(iterator, &physical_mappings)
+	{
+		struct active_physical_mapping *mapping = CONTAINER_OF(
+			iterator, struct active_physical_mapping, node);
+
+		if (is_range_in_range(ptr, alloc_size,
+			mapping->mapping_ptr,
+			mapping->mapping_size))	{
+			space_already_mapped = true;
+		}
+	}
+
+	/* If the space was not already mapped we map physical pages */
+	if (!space_already_mapped) {
+		void *aligned_ptr = (void *)ALIGN_DOWN((uintptr_t)ptr,
+			CONFIG_MM_DRV_PAGE_SIZE);
+		size_t aligned_alloc_size =
+			ALIGN_UP((uintptr_t)((uintptr_t)ptr + alloc_size),
+				CONFIG_MM_DRV_PAGE_SIZE) - (uintptr_t)aligned_ptr;
+
+		sys_mm_drv_map_region(aligned_ptr, (uintptr_t)NULL,
+			aligned_alloc_size, SYS_MM_MEM_PERM_RW);
+		struct active_physical_mapping *new_mapping = rzalloc(SOF_MEM_ZONE_SYS,
+			0, SOF_MEM_CAPS_RAM, sizeof(struct active_physical_mapping));
+
+		if (new_mapping == NULL) {
+			rfree(new_allocation);
+			return NULL;
+		}
+		new_mapping->mapping_ptr = aligned_ptr;
+		new_mapping->mapping_size = aligned_alloc_size;
+		list_item_append(&physical_mappings, &new_mapping->node);
+	}
+
+	list_item_append(&heap->active_allocations, &new_allocation->node);
+
+	return ptr;
+}
+
+
+/**
+ * @brief Free virtual memory heap
+ *
+ * Free the virtual memory heap object and its child allocations
+ * @param heap Ptr to the heap that is supposed to be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free_heap(struct virtual_memory_heap *heap)
+{
+	if (!list_is_empty(&heap->active_allocations))
+		return -ENOTEMPTY;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		if (heap->physical_blocks_allocators[i] != 0) {
+			rfree(heap->physical_blocks_allocators[i]->bitmap->bundles);
+			rfree(heap->physical_blocks_allocators[i]->bitmap);
+			rfree(heap->physical_blocks_allocators[i]);
+		}
+	}
+	rfree(heap);
+	return 0;
+}
+
+/**
+ * @brief Free ptr allocated on given heap
+ *
+ * Free the ptr allocation. After free action is complete
+ * check if any physical memory block can be freed up as
+ * result.
+ * @param heap Ptr to the heap that is supposed to be freed
+ * @param ptr pointer that should be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free(struct virtual_memory_heap *heap, void *ptr)
+{
+	struct list_item *iterator;
+	struct active_allocation *allocation = NULL;
+	int retval;
+
+	list_for_item(iterator, &heap->active_allocations)
```

Comment:
Some of the code blocks are large/complex and could do with more inlien comments.